SharkCage/CageManager/CageManager.cpp

```diff
@@ -0,0 +1,248 @@
+	DWORD cch_domain_name = 0;
+	SID_NAME_USE sid_type;
+	DWORD sid_buffer_size = INITIAL_SIZE;
+
+	// First call of the function in order to get the size needed for the SID
+	::LookupAccountName(
+		NULL,            // Computer name. NULL for the local computer  
+		group_name_buf.data(),
+		NULL,
+		&cb_sid,
+		domain_name.data(),
+		&cch_domain_name,
+		&sid_type
+	);
+
+	// Second call of the function in order to get the SID
+	std::unique_ptr<PSID, decltype(sid_deleter)> sid((PSID*)::LocalAlloc(LPTR, cb_sid), sid_deleter);
+	
+	::LookupAccountName(
+		NULL,
+		group_name_buf.data(),
+		sid.get(),
+		&cb_sid,
+		domain_name.data(),
+		&cch_domain_name,
+		&sid_type
+	);
+
+	return sid;
+}
+
+// FIXME: move this and other duplicate functions and their respective implementation in cage service to its own helper class or something like that
+bool BeginsWith(const std::wstring &string, const std::wstring &prefix)
+{
+	if (prefix.length() > string.length())
+	{
+		return false;
+		// Throw Exception "Bad parameters: prefix longer than the actual string"
+	}
+	else
+	{
+		if (string.compare(0, prefix.length(), prefix) == 0)
+		{
+			return true;
+		}
+		else
+		{
+			return false;
+		}
+	}
+}
+
+// Function to decode the message and do a respective action
+// "START_PC" "path/to.exe"
+std::wstring OnReceive(const std::wstring &message)
+{
+	std::wstring path = L"";
+	if (BeginsWith(message, ManagerMessageToString(ManagerMessage::START_PROCESS)))
+	{
+		// Start process
+		path = message.substr(9);
+	}
+	else if (BeginsWith(message, ManagerMessageToString(ManagerMessage::STOP_PROCESS)))
+	{
+		// Stop process
+	}
+	else
+	{
+		std::wcout << "Received unrecognized message: " << message << std::endl;
+	}
+	return path;
+}
+
+bool CreateACL(std::unique_ptr<PSID, decltype(sid_deleter)> group_sid)
+{
+	DWORD result;
+	PACL acl = NULL;
+	PSECURITY_DESCRIPTOR security_descriptor = NULL;
+	SECURITY_ATTRIBUTES security_attributes;
+	HDESK new_desktop = NULL;
+	//Listen for the message
+	std::wstring message = network_manager.Listen();
+	std::wstring path = OnReceive(message);
+
+	// create SID for BUILTIN\Administrators group
+	PSID sid_admin = NULL;
+	SID_IDENTIFIER_AUTHORITY sid_authnt = SECURITY_NT_AUTHORITY;
+	if (!::AllocateAndInitializeSid(&sid_authnt, 2, SECURITY_BUILTIN_DOMAIN_RID, DOMAIN_ALIAS_RID_ADMINS, 0, 0, 0, 0, 0, 0, &sid_admin))
+	{
+		std::cout << "Obtain Admin SID Error: " << ::GetLastError() << std::endl;
+		return false;
+	}
+
+	// create EXPLICIT_ACCESS structure for an ACE
+	EXPLICIT_ACCESS explicit_access_group = { 0 };
+	EXPLICIT_ACCESS explicit_access_admin = { 0 };
+
+	// EXPLICIT_ACCESS for created groupc 
+	explicit_access_group.grfAccessPermissions = GENERIC_ALL;
+	explicit_access_group.grfAccessMode = SET_ACCESS;
+	explicit_access_group.grfInheritance = NO_INHERITANCE;
+	explicit_access_group.Trustee.TrusteeForm = TRUSTEE_IS_SID;
+	explicit_access_group.Trustee.TrusteeType = TRUSTEE_IS_GROUP;
+	
+	wchar_t *group_sid_string;
```

Comment:
after looking at this again: this should be a unique_ptr, the `::LocalFree` after the assignment frees memory which is used by the `EXPLICIT_ACCESS` struct